### ISSUE TITLE
Add rule for reference style links with no `[]` suffix

### ISSUE THREAD

**Issue by ChrisBAshton** [markdownlint/markdownlint#351]
## New Rule Checklist

Before suggesting a rule for inclusion please make sure your suggestion meets these criteria for rule built into markdownlint:
 - [x] allows a user to lint for a specific syntax divergence across the multiple flavors and styles of markdown
 - [x] does not dictate any one specific style, but enables an end user to enable, disable, or configure the specific style of enforcement desired

## Describe The Rule:
Consider the following markdown:

```md
Download [Foo].

Download [Foo][].
```

Both links are valid, and refer to a reference further down in the document:

```md
[Foo]: https://example.com
```

We should be able to enforce one style of reference link: either `[Foo]` **or** `[Foo][]`. By default, we should simply allow both.

## Why Should This Be Included In Markdownlint?:

Stylistically: some users may think the `[Foo]` example is more readable, and opt to enforce that style. Others may prefer the more explicit nature of `[Foo][]`.

Functionally, the `[Foo]` style can be ambiguous, because of a third style: `[Foo][alias]`. It seems that [space is allowed between the sets of brackets](https://www.markdownguide.org/basic-syntax/#formatting-the-first-part-of-the-link), i.e. `[Foo][alias]` can also be represented as `[Foo] [alias]`. Without looking elsewhere in the document, it's impossible to know if `[Foo] [alias]` is one link or two.

With this input:

```md
[alias]: https://www.google.com
```

We'd get this output:

```html
<a href="https://www.google.com">Foo</a>.
```

And with this input:

```md
[Foo]: https://www.example.com
[alias]: https://www.google.com
```

We'd get this output:

```html
<a href="https://www.example.com">Foo</a> <a href="https://www.google.com">alias</a>.
```

For this reason, people may want to enforce the more explicit `[Foo][]` style, such that `[Foo]` raises an error.

Finally, some markdown parsers can be quite buggy when parsing this shortcut reference link style. In our case, we're affected by [this issue](https://github.com/vmg/redcarpet/issues/471), whereby content in parentheses is interpreted as part of the link:

```markdown
See [foo] (example: bla)

[foo]: /bar.html
```

(Being parsed as `See <a href="(example: bla)">foo<a>`).

## Can We Help You Implement This?:

I'm happy to give this a go. Let me know what rule number this should be, and ideally point me to a good reference PR where other rules have been added!
